// thorjs-tx/fields.js
module.exports = function getFields() {
  return [
    {
      name: 'ChainTag',
      length: 1
    },
    {
      name: 'BlockRef',
      length: 8,
      allowZero: true,
      allowLess: true,
      default: new Buffer([])
    },
    {
      name: 'Expiration',
      length: 4,
      allowLess: true,
      default: new Buffer([])
    },
    {
      name: 'Clauses',
      callback(Clauses) {
        let result = []
        let to = 0
        let value = 1
        let data = 2
        if (Clauses && Clauses.length) {
          result = Clauses.map(item => {
            let clausesBody = []
            clausesBody[to] = item.to
            clausesBody[value] = item.value
            clausesBody[data] = item.data || ''
            return clausesBody
          })
        }

        return result
      }
    },
    {
      name: 'GasPriceCoef',
      length: 1,
      allowLess: true,
      default: new Buffer([])
    },
    {
      name: 'Gas',
      length: 8,
      allowLess: true,
      default: new Buffer([])
    },
    {
      name: 'DependsOn',
      allowless: true,
      length: 32,
      default: new Buffer([])
    },
    {
      name: 'Nonce',
      length: 8,
      allowLess: true,
      allowZero: true,
      default: new Buffer([])
    },
    {
      name: 'Reserved',
      default: []
    }
  ]
}
