// thorify/dist/extend/accounts.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Tx = require("thorjs-tx");
const debug = require("debug")("thor:injector");
const EthLib = require("eth-lib/lib");
const utils = require("../utils");
const extendAccounts = function (web3) {
    const proto = Object.getPrototypeOf(web3.eth.accounts);
    // signTransaction supports both callback and promise style
    proto.signTransaction = function signTransaction(tx, privateKey, callback) {
        debug("tx to sign: %O", tx);
        const thorTx = utils.ethToThorTx(tx);
        const sign = function (tx) {
            return __awaiter(this, void 0, void 0, function* () {
                if (!tx.chainTag) {
                    const chainTag = yield web3.eth.getChainTag();
                    if (chainTag) {
                        tx.chainTag = chainTag;
                    }
                    else {
                        throw new Error("error getting chainTag");
                    }
                }
                if (!tx.blockRef) {
                    const blockRef = yield web3.eth.getBlockRef();
                    if (blockRef) {
                        tx.blockRef = blockRef;
                    }
                    else {
                        throw new Error("error getting blockRef");
                    }
                }
                if (!tx.gas) {
                    const gas = yield web3.eth.estimateGas({
                        from: EthLib.account.fromPrivate(utils.toPrefixedHex(privateKey)).address,
                        to: tx.clauses.length ? tx.clauses[0].to : "",
                        value: tx.clauses.length ? tx.clauses[0].value : 0,
                        data: tx.clauses.length ? tx.clauses[0].data : "0x",
                    });
                    if (gas) {
                        tx.gas = gas;
                    }
                    else {
                        throw new Error("error getting gas");
                    }
                }
                if (!tx.nonce) {
                    tx.nonce = utils.newNonce();
                }
                debug(tx);
                const thorTx = Tx({
                    ChainTag: tx.chainTag,
                    BlockRef: tx.blockRef,
                    Expiration: tx.expiration,
                    Clauses: tx.clauses,
                    GasPriceCoef: tx.gasPriceCoef,
                    Gas: tx.gas,
                    DependsOn: tx.dependsOn,
                    Nonce: tx.nonce,
                });
                thorTx.sign(utils.sanitizeHex(privateKey));
                const rawTx = thorTx.serialize();
                const result = {
                    rawTransaction: utils.toPrefixedHex(rawTx.toString("hex")),
                    messageHash: utils.toPrefixedHex(thorTx.signatureHash()),
                };
                return result;
            });
        };
        // for supporting both callback and promise
        if (callback instanceof Function) {
            sign(thorTx).then((ret) => {
                return callback(null, ret);
            }).catch((e) => {
                return callback(e);
            });
        }
        else {
            return sign(thorTx);
        }
    };
    proto.recoverTransaction = function recoverTransaction(encodedRawTx) {
        const values = EthLib.RLP.decode(encodedRawTx);
        const signingDataHex = EthLib.RLP.encode(values.slice(0, 9));
        const signingHash = utils.hash(Buffer.from(utils.sanitizeHex(signingDataHex), "hex"));
        const signature = values[9];
        const singingHashBuffer = Buffer.from(utils.sanitizeHex(signingHash), "hex");
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), "hex");
        const address = utils.recover(singingHashBuffer, signatureBuffer);
        return address;
    };
    proto.hashMessage = function hashMessage(data) {
        const message = web3.extend.utils.isHexStrict(data) ? web3.extend.utils.hexToBytes(data) : data;
        const messageBuffer = Buffer.from(message);
        return utils.hash(messageBuffer);
    };
    proto.sign = function sign(data, privateKey) {
        const hash = this.hashMessage(data);
        const hashBuffer = Buffer.from(utils.sanitizeHex(hash), "hex");
        const privateKeyBuffer = Buffer.from(utils.sanitizeHex(privateKey), "hex");
        const signature = utils.sign(hashBuffer, privateKeyBuffer);
        return {
            message: data,
            messageHash: utils.toPrefixedHex(hash),
            signature: utils.toPrefixedHex(signature),
        };
    };
    proto.recover = function recover(message, signature, preFixed) {
        const args = [].slice.apply(arguments);
        if (web3.extend.utils._.isObject(message)) {
            return this.recover(message.messageHash, message.signature, true);
        }
        if (!preFixed) {
            message = this.hashMessage(message);
        }
        const hexBuffer = Buffer.from(utils.sanitizeHex(message), "hex");
        const signatureBuffer = Buffer.from(utils.sanitizeHex(signature), "hex");
        return utils.recover(hexBuffer, signatureBuffer);
    };
};
exports.extendAccounts = extendAccounts;
