// thor-kit/dist/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Bytes {
    constructor(buf) {
        this.buffer = buf;
    }
    static from(hex, prefix = '0x') {
        if (!hex.startsWith(prefix))
            throw new Error('prefix mismatch');
        hex = hex.substr(prefix.length);
        if (!Bytes.isHex.test(hex))
            throw new Error('not hex format');
        if (hex.length % 2 != 0) {
            throw new Error('odd hex');
        }
        return new Bytes(new Buffer(hex, 'hex'));
    }
    toString(prefix = '0x') {
        return (prefix || '') + this.buffer.toString('hex');
    }
    trimLeft(v = 0) {
        let i = 0;
        for (; i < this.buffer.length; i++) {
            if (this.buffer[i] != v)
                break;
        }
        return new Bytes(this.buffer.slice(i));
    }
}
Bytes.isHex = /^[0-9a-f]*$/i;
exports.Bytes = Bytes;
class Address {
    static from(hex, prefix = '0x') {
        let bytes = Bytes.from(hex, prefix);
        if (bytes.buffer.length != 20)
            throw new Error("address should be 20 bytes");
        return new Address(bytes.buffer);
    }
    constructor(buf) {
        let newBuf = new Buffer(20);
        buf.copy(newBuf, newBuf.length > buf.length ? newBuf.length - buf.length : 0, buf.length > newBuf.length ? buf.length - newBuf.length : 0);
        this.buffer = newBuf;
    }
    toString(prefix = '0x') {
        return (prefix || '') + this.buffer.toString('hex');
    }
}
exports.Address = Address;
class Bytes32 {
    static from(hex, prefix = '0x') {
        let bytes = Bytes.from(hex, prefix);
        if (bytes.buffer.length != 32)
            throw new Error("bytes32 should be 32 bytes");
        return new Bytes32(bytes.buffer);
    }
    constructor(buf) {
        let newBuf = new Buffer(32);
        buf.copy(newBuf, newBuf.length > buf.length ? newBuf.length - buf.length : 0, buf.length > newBuf.length ? buf.length - newBuf.length : 0);
        this.buffer = newBuf;
    }
    toString(prefix = '0x') {
        return (prefix || '') + this.buffer.toString('hex');
    }
}
exports.Bytes32 = Bytes32;
const bignumber_js_1 = require("bignumber.js");
class BigInt {
    static from(v) {
        let bn;
        if (typeof v === 'string') {
            bn = new bignumber_js_1.BigNumber(v);
        }
        else if (typeof v === 'number') {
            bn = new bignumber_js_1.BigNumber(v);
        }
        else {
            bn = v;
        }
        if (!bn.isInteger()) {
            throw new Error('not a integer');
        }
        let hex = bn.toString(16);
        if (hex.length % 2 != 0)
            hex = '0' + hex;
        return new BigInt(new Buffer(hex, 'hex'));
    }
    constructor(buf) {
        this.buffer = new Bytes(buf).trimLeft().buffer;
    }
    toBigNumber() {
        return new bignumber_js_1.BigNumber(this.buffer.toString("hex"), 16);
    }
    toString(base = 16) {
        let prefix = (base == 16) ? '0x' : '';
        return prefix + this.toBigNumber().toString(base);
    }
}
exports.BigInt = BigInt;
